src/hotspot/share/gc/shared/partialArrayState.hpp

```diff
@@ -91,47 +95,136 @@ class PartialArrayState {
   ~PartialArrayStateAllocator();
 
   NONCOPYABLE(PartialArrayStateAllocator);
 
   // Create a new state, obtaining the memory for it from the free-list or
-  // arena associated with worker_id.
-  PartialArrayState* allocate(uint worker_id,
-                              oop src, oop dst,
+  // from the associated manager.
+  PartialArrayState* allocate(oop src, oop dst,
                               size_t index, size_t length,
                               size_t initial_refcount);
 
   // Decrement the state's refcount.  If the new refcount is zero, add the
   // state to the free-list associated with worker_id.  The state must have
   // been allocated by this allocator, but that allocation doesn't need to
   // have been associated with worker_id.
-  void release(uint worker_id, PartialArrayState* state);
+  void release(PartialArrayState* state);
+};
+
+// This class provides memory management for PartialArrayStates.
+//
+// States are allocated using an allocator object. Those allocators in turn
+// may request memory for a state from their associated manager. The manager
+// is responsible for obtaining and releasing memory used for states by the
+// associated allocators.
+//
+// A state may be allocated by one allocator, but end up on the free-list of a
+// different allocator.  This can happen because a task referring to the state
+// may be stolen from the queue where it was initially added.  This is permitted
+// because a state's memory won't be reclaimed until all of the allocators
+// associated with the manager that is ultimately providing the memory have
+// been deleted and the manager is reset.
+//
+// A manager is used in two distinct and non-overlapping phases.
+//
+// - allocating: This is the initial phase.  During this phase, new allocators
+// may be created, and allocators may request memory from the manager.
+//
+// - releasing: When an allocator is destroyed the manager transitions to this
+// phase.  It remains in this phase until all extent allocators associated with
+// this manager have been destroyed.  During this phase, new allocators may not
+// be created, nor may extent allocators request memory from this manager.
+//
+// Once all the associated allocators have been destroyed the releasing phase
+// ends and the manager may be reset or deleted.  Resetting transitions back
+// to the allocating phase.
+class PartialArrayStateManager : public CHeapObj<mtGC> {
+  friend class PartialArrayStateAllocator;
+
+  // Use an arena for each allocator, for thread-safe concurrent allocation by
+  // different allocators.
+  Arena* _arenas;
+
+  // Limit on the number of allocators this manager supports.
+  uint _max_allocators;
+
+  // CounterValues is an integral type large enough to encode a pair of
+  // allocator counters as a single unit for atomic manipulation.
+  using CounterValues = LP64_ONLY(uint64_t) NOT_LP64(uint32_t);
+  using Counter = LP64_ONLY(uint32_t) NOT_LP64(uint16_t);
```

Comment:
Given the max-value has type `uint`, using the larger type on both 32/64 bit systems should be simpler and it should not cause any noticeable perf regression, since registering/releasing allocators should be infrequent. WDYT?

Reply:
I assumed 16bits of worker threads was quite sufficient for a 32bit platform.
And I misremembered and thought 32bit platforms couldn't be relied upon for a
64bit atomic add and maybe other 64bit operations. And this code is definitely
not super performance critical. So yeah, I could drop the platform-conditional
definition of Counter. I don't think it makes much difference to the code.  I
guess the type aliases could be dropped and just use bare uint32/64_t.  Not
sure that's actually an improvement.

Reply:
I think unifying 32 and 64 bit system is an improvement -- being able to reason with concrete types. As for type aliases, it's rather subjective; I find `uint*_t` more familiar, but up to  you.